core/client/src/main/java/alluxio/client/file/FileInStream.java

```diff
@@ -416,78 +351,116 @@ private long getCurrentBlockId() {
   }
 
   /**
-   * @param pos the position to get the block size for
-   * @return the size of the current block
-   */
-  protected long getBlockSize(long pos) {
-    // The size of the last block, 0 if it is equal to the normal block size
-    long lastBlockSize = mFileLength % mBlockSize;
-    // If we are not in the last block or if the last block is equal to the normal block size,
-    // return the normal block size. Otherwise return the block size of the last block.
-    if (mFileLength - pos > lastBlockSize) {
-      return mBlockSize;
-    } else {
-      return lastBlockSize;
-    }
-  }
-
-  /**
    * Logs IO exceptions thrown in response to the worker cache request. If the exception is not an
    * expected exception, a warning will be logged with the stack trace.
    */
-  private void logCacheStreamIOException(IOException e) {
+  private void handleCacheStreamIOException(IOException e) throws IOException {
     if (e.getCause() instanceof BlockAlreadyExistsException) {
       LOG.warn(BLOCK_ID_EXISTS_SO_NOT_CACHED, getCurrentBlockId());
     } else {
       LOG.warn(BLOCK_ID_NOT_CACHED, getCurrentBlockId(), e);
     }
+    closeOrCancelCacheStream();
   }
 
   /**
-   * Similar to {@link #checkAndAdvanceBlockInStream()}, but a specific position can be specified
-   * and the stream pointer will be at that offset after this method completes.
+   * Only updates {@link #mCurrentCacheStream}, {@link #mCurrentBlockInStream} and
+   * {@link #mStreamBlockId} to be in-sync the current block (i.e. {@link #getCurrentBlockId()}).
+   * This function can be called multiple times without side effect. It is recommended to be
```

Comment:
yes to the first question. I have reworded my comment. it is only recommended because if the block stream doesn't change, then this is not required.